mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+        }
+      }
+      networks[filteredIndex++] = network;
+    }
+    return Arrays.copyOf(networks, filteredIndex);
+  }
+
+  /**
+   * Returns network handle of device's current default connected network used for
+   * communication.
+   * Returns -1 when not implemented.
+   */
+  public long getDefaultNetId() {
+    Network network = getDefaultNetwork();
+    return network == null ? -1 : network.getNetworkHandle();
+  }
+
+  /** Returns the current default {@link Network}, or {@code null} if disconnected. */
+  private Network getDefaultNetwork() {
+    Network defaultNetwork = null;
+    defaultNetwork = mConnectivityManager.getActiveNetwork();
+    if (defaultNetwork != null) {
+      return defaultNetwork;
+    }
+    // getActiveNetwork() returning null cannot be trusted to indicate disconnected
+    // as it suffers from https://crbug.com/677365.
+    // Check another API to return the NetworkInfo for the default network. To
+    // determine the default network one can find the network with
+    // type matching that of the default network.
+    final NetworkInfo defaultNetworkInfo = mConnectivityManager.getActiveNetworkInfo();
+    if (defaultNetworkInfo == null) {
+      return null;
+    }
+    final Network[] networks = getAllNetworksFiltered(null);
+    for (Network network : networks) {
+      final NetworkInfo networkInfo = getRawNetworkInfo(network);
+      if (networkInfo != null &&
+          (networkInfo.getType() == defaultNetworkInfo.getType()
+           // getActiveNetworkInfo() will not return TYPE_VPN types due to
+           // https://android.googlesource.com/platform/frameworks/base/+/d6a7980d
+           // so networkInfo.getType() can't be matched against
+           // defaultNetworkInfo.getType() but networkInfo.getType() should
+           // be TYPE_VPN. In the case of a VPN, getAllNetworks() will have
+           // returned just this VPN if it applies.
+           || networkInfo.getType() == TYPE_VPN)) {
+        // Android 10+ devices occasionally return multiple networks
+        // of the same type that are stuck in the CONNECTING state.
+        // Now that Java asserts are enabled, ignore these zombie
+        // networks here to avoid hitting the assert below. crbug.com/1361170
+        if (defaultNetwork != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+          // If `network` is CONNECTING, ignore it.
+          if (networkInfo.getDetailedState() == NetworkInfo.DetailedState.CONNECTING) {
+            continue;
+          }
+          // If `defaultNetwork` is CONNECTING, ignore it.
+          NetworkInfo prevDefaultNetworkInfo = getRawNetworkInfo(defaultNetwork);
+          if (prevDefaultNetworkInfo != null &&
+              prevDefaultNetworkInfo.getDetailedState() == NetworkInfo.DetailedState.CONNECTING) {
+            defaultNetwork = null;
+          }
+        }
+        if (defaultNetwork != null) {
+          // TODO(crbug.com/40060873): Investigate why there are multiple
```

Comment:
crbug

Another note: we'll probably want to track these bugs and it fixes get put into Chromium code, we can port it here has well